frontend/src/services/GooglePlacesService.ts

```diff
@@ -0,0 +1,183 @@
+import axios, { AxiosError, AxiosResponse } from "axios";
+import * as dotenv from "dotenv";
+import { RestaurantDetails } from ".././types/RestaurantDetails";
+
+/* eslint-disable  no-console */
+// todo: remove this eslint-disable
+class GooglePlacesService {
+  public currRestaurants: Map<string, RestaurantDetails>;
+
+  constructor() {
+    dotenv.config();
+    this.currRestaurants = new Map();
+  }
+
+  getRestaurantPlaceID(
+    restaurantName: string,
+    restaurantLocation: string
+  ): Promise<string> {
+    const dataURI: string = encodeURIComponent(restaurantName);
+    let responseData: any[];
+
+    return axios
+      .get(
+        "https://maps.googleapis.com/maps/api/place/findplacefromtext/json?",
+        {
+          params: {
+            input: dataURI,
+            inputtype: "textquery",
+            fields: "place_id",
+            key: process.env.REACT_APP_GOOGLE_PLACES_API_KEY,
+          },
+        }
+      )
+      .then((response: AxiosResponse) => {
+        if (response.data.candidates.length === 1) {
+          return Promise.resolve(response.data.candidates?.[0].place_id);
+        }
+        const promises: Array<Promise<boolean>> = [];
+        responseData = response.data.candidates;
+
+        for (const currPlace of response.data.candidates) {
+          promises.push(
+            this.isCorrectRestaurantLocation(
+              restaurantLocation,
+              currPlace.place_id
+            )
+          );
+        }
+
+        return Promise.all(promises);
+      })
+      .then((result) => {
+        if (!Array.isArray(result)) {
+          return Promise.resolve(result);
+        }
+
+        for (let i = 0; i < result.length; i++) {
+          if (result[i]) {
+            return Promise.resolve(responseData[i].place_id);
+          }
+        }
+
+        return Promise.reject("No restaurant location matched search");
+      })
+      .catch((err: AxiosError) => {
+        //Google API error message
+        console.log(err?.response?.data?.error);
+
+        //Axios entire error message
+        console.log(err);
+
+        return Promise.reject(err);
+      });
+  }
+
+  isCorrectRestaurantLocation(
+    restaurantLocation: string,
+    currPlaceID: string
+  ): Promise<boolean> {
+    return axios
+      .get("https://maps.googleapis.com/maps/api/place/details/json?", {
+        params: {
+          place_id: currPlaceID,
+          fields: "formatted_address",
+          key: process.env.REACT_APP_GOOGLE_PLACES_API_KEY,
+        },
+      })
+      .then((response: AxiosResponse) => {
+        const currRestaurantAddress = response?.data?.result?.formatted_address;
+        if (currRestaurantAddress.includes(restaurantLocation)) {
```

Comment:
Hmm how do we know that our restaurant location will match to `currRestaurantAddress`? Using string operations is fragile imo because we can't guarantee that it will always match, there may be a lot of edge cases.

Reply:
Yeah that's true, I see where you're coming from. I couldn't think of another way to verify if the current restaurant is the one that the user is searching for though- and also with correspondence of the fact that we're storing the user-submitted locations in the DB, I think _for now until we figure out a better way to handle both situations_, we can add extra notes in the promotion submission page that tells the user to make sure they're inserting the address correctly (or ask for particular fields like Address Line 1, Address Line 2, City, Province, Postal Code)?